src/poh_service.rs

```diff
@@ -46,55 +44,74 @@ impl PohService {
     }
 
     pub fn new(
-        bank: Arc<Bank>,
-        sender: Sender<Vec<Entry>>,
         poh_recorder: PohRecorder,
         config: PohServiceConfig,
         poh_exit: Arc<AtomicBool>,
-    ) -> Self {
+    ) -> (Self, Sender<WorkingLeader>) {
         // PohService is a headless producer, so when it exits it should notify the banking stage.
         // Since channel are not used to talk between these threads an AtomicBool is used as a
         // signal.
         let poh_exit_ = poh_exit.clone();
+        let (leader_sender, leader_receiver) = channel();
         // Single thread to generate ticks
         let tick_producer = Builder::new()
             .name("solana-poh-service-tick_producer".to_string())
             .spawn(move || {
-                let mut poh_recorder_ = poh_recorder;
-                let sender = sender.clone();
-                let bank = bank.clone();
+                let mut poh_recorder = poh_recorder;
+                let leader_receiver = leader_receiver;
                 let return_value =
-                    Self::tick_producer(&bank, &sender, &mut poh_recorder_, config, &poh_exit_);
+                    Self::tick_producer(&leader_receiver, &mut poh_recorder, config, &poh_exit_);
                 poh_exit_.store(true, Ordering::Relaxed);
                 return_value
             })
             .unwrap();
 
-        Self {
-            tick_producer,
-            poh_exit,
-        }
+        (
+            Self {
+                tick_producer,
+                poh_exit,
+            },
+            leader_sender,
+        )
     }
 
     fn tick_producer(
-        bank: &Arc<Bank>,
-        sender: &Sender<Vec<Entry>>,
+        leader_receiver: &Receiver<WorkingLeader>,
         poh: &mut PohRecorder,
         config: PohServiceConfig,
         poh_exit: &AtomicBool,
     ) -> Result<()> {
+        let mut leader = None;
         loop {
+            if leader.is_none() {
+                let e = leader_receiver.try_recv();
+                leader = match e {
+                    Err(TryRecvError::Empty) => None,
+                    _ => Some(e?),
+                };
+            }
             match config {
                 PohServiceConfig::Tick(num) => {
                     for _ in 1..num {
-                        poh.hash()?;
+                        poh.hash();
                     }
                 }
                 PohServiceConfig::Sleep(duration) => {
                     sleep(duration);
                 }
             }
-            poh.tick(&bank, sender)?;
+            let e = if let Some(ref current_leader) = leader {
```

Comment:
No one-letter variables outside of one-line closures please. You're using `e` for error, I presume, but this is a `Result`.